src/main/scala/mesosphere/marathon/util/LifeCycledCloseable.scala

```diff
@@ -0,0 +1,39 @@
+package mesosphere.marathon
+package util
+
+import java.io.Closeable
+import scala.util.Try
+
+trait LifeCycledCloseableLike[T <: Closeable] extends Closeable {
+  def closeable: T
+  def beforeClose(fn: () => Unit): Unit
+  def removeBeforeClose(fn: () => Unit): Unit
+}
+
+class LifeCycledCloseable[T <: Closeable](val closeable: T) extends LifeCycledCloseableLike[T] {
+  @volatile private[this] var closed: Boolean = false
+  @volatile private[this] var beforeCloseHooks: List[() => Unit] = Nil
+
+  override def close(): Unit = {
+    val hooks = synchronized {
+      closed = true
+      val oldHooks = beforeCloseHooks
+      beforeCloseHooks = Nil
+      oldHooks
+    }
+    hooks.foreach { hook =>
+      Try(hook())
+    }
+    closeable.close()
+    beforeCloseHooks = Nil
```

Comment:
second Nil-ing outside of `synchronized` block?

Reply:
Yeah, good catch; this is not necessary.